File: Sources/termfarm/Persistence.swift
import Foundation

func saveURL() -> URL {
    let home = FileManager.default.homeDirectoryForCurrentUser
    let dir  = home.appendingPathComponent(".local/share/termfarm", isDirectory: true)

    try? FileManager.default.createDirectory(
        at: dir,
        withIntermediateDirectories: true
    )

    return dir.appendingPathComponent("save.json")
}

func loadFarm() throws -> FarmState {
    let url = saveURL()
    let data = try Data(contentsOf: url)
    var farm = try JSONDecoder().decode(FarmState.self, from: data)

    updateMarketIfNeeded(farm: &farm)

    return farm
}

func saveFarm(_ farm: FarmState) throws {
    let data = try JSONEncoder().encode(farm)
    try data.write(to: saveURL(), options: .atomic)
}

File: src/persistence.rs
use crate::{market::update_market_if_needed, models::FarmState};
use std::fs;

pub fn save_path() -> String {
    let home = dirs::home_dir()
        .map(|p| p.to_string_lossy().into_owned())
        .unwrap_or_default();
    let dir = home + "/.local/share/termfarm";

    _ = fs::create_dir_all(&dir);

    dir + "/save.json"
}

pub fn load_farm() -> FarmState {
    let path = save_path();
    let mut farm = fs::read_to_string(path)
        .ok()
        .and_then(|s| serde_json::from_str(&s).ok())
        .unwrap();

    update_market_if_needed(&mut farm);

    farm
}

pub fn save_farm(farm: &FarmState) -> bool {
    match serde_json::to_string_pretty(farm) {
        Ok(json) => fs::write(save_path(), json).is_ok(),
        Err(_) => false,
    }
}
